### PR TITLE
pkgs: add consideration for new packages

### PR DESCRIPTION
**Pull request by zimbatm**
## Description of changes

Add a light decision-making framework to help decide what fits in the project or not. To avoid nixpkgs becoming a dump of personal projects.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/README.md
Comment (mweinelt) [inline pkgs/README.md:83]: I'm hesitant to characterize `license` as optional.

File: pkgs/README.md
Comment (mweinelt) [inline pkgs/README.md:30]: Wondering if we should specify a timeframe here, like

```suggestion
* Are you going to keep maintaining the package at least 6 months from today?
```

**Review by 7c6f434c — DISMISSED** [review]
For Nixpkgs that had releases _less_ stable than the average state of unstable for a few years to pretend that not having versioned releases is always a sign of instability is both disconnected from reality and hypocritical

**Review by andir — APPROVED** [review]

**Review by matthiasbeyer — APPROVED** [review]

File: pkgs/README.md
Comment (AndersonTorres) [inline pkgs/README.md:83]: platforms and mainProgram should be set, too.

Comment (zimbatm) [reply pkgs/README.md:83]: Personally I'm also expecting the license to be set, but let's have that conversation separately ok?

**Comment by zimbatm** [comment]
> For Nixpkgs that had releases less stable than the average state of unstable for a few years to pretend that not having versioned releases is always a sign of instability is both disconnected from reality and hypocritical

Was the personal attack necessary? I agree that there are some cases where packaging an unreleased version makes sense, but generally speaking, it's a good signal that the project is not ready to be used. I could change the wording to leave more room for interpretation or add some explicit exceptions.

File: pkgs/README.md
Comment (7c6f434c) [reply pkgs/README.md:83]: Requiring license is probably defensible from «checking it is supposed to be packageable» point of view.

File: pkgs/README.md
Comment (zimbatm) [reply pkgs/README.md:30]: It's a good question, I don't know.

On the surface, it seems reasonable. But it also opens up new questions like what is the appropriate duration, is there a single duration that works for all packages, and what happens if they don't respect that commitment?

Or we ask the contributor to make a good-faith commitment and leave it unspecified.

**Comment by 7c6f434c** [comment]
> Was the personal attack necessary?

Was it really the attack? I think we should be more careful not to lose at least the implications of our own experience, as we don't always have access to interpret fully the experience of other projects.

> generally speaking, it's a good signal that the project is not ready to be used

Given that the packages I trust most to keep working are often the no-versioned-releases-ever ones, if we want to ask for some property, we should _start_ with stating the property and not just put a pretty unreliable proxy in its place.

File: pkgs/README.md
Comment (zimbatm) [inline pkgs/README.md:28]: @7c6f434c what better metric for maturity do you propose? Let's restart the convo in this thread.

Comment (zimbatm) [reply pkgs/README.md:28]: Also what kind of packages do you have in mind? The vast majority of packages have releases.

Comment (7c6f434c) [reply pkgs/README.md:28]: I don't think there is a good universal metric, and I don't think there can be. 

If what we care about is whether the package ready for general use, we should start by asking «Is the package ready for general use?» Sure, nothing wrong with adding «For example, if the author(s) is/are working towards the first release, it is probably too early.»

Quicklisp for one example is half-full of packages that don't bother with versions and just publish commits fixing whatever small bugs are discovered.

File: pkgs/README.md
Comment (mweinelt) [reply pkgs/README.md:30]: A more reasonable timeframe would be the next NixOS release cycle, but I'm fine with a good faith commitment as well.

File: pkgs/README.md
Comment (zimbatm) [reply pkgs/README.md]: Thanks, that's useful. Adopted.

Comment (7c6f434c) [reply pkgs/README.md]: Thanks!

File: pkgs/README.md
Comment (zimbatm) [reply pkgs/README.md:31]: It would be nice to move towards that for sure. I wonder what existing maintainers would think of this. Keeping this open for now to let it simmer a bit.

File: pkgs/README.md
Comment (RaitoBezarius) [inline pkgs/README.md:28]: Remark: I will go further, I had discussions with @Foxboron on this who had some of his projects "too early packaged" and he was not really comfortable with this.

For example, we can also be explicit and say "if the author explicitly asks for not being packaged", etc. I mean, it's obvious in retrospective in packaging `unstable-XYZ`-type of packages.

I would just like to distinguish:

- software that is used and can be used because it has been there for a while, but no release or security fixes in latest HEAD, etc.
- software that is being actively developed and can be already ~usable as a sneak peek, but with absolute no quality guarantees.

The former is fine without "first release", the latter is very much meh IMHO.

File: pkgs/README.md
Comment (RaitoBezarius) [inline pkgs/README.md:33]: Offer alternatives here IMHO: "consider a NUR, etc."

**Review by RaitoBezarius — APPROVED** [review]

**Comment by RaitoBezarius** [comment]
Giving examples may be interesting, not sure how feasible this may be.

File: pkgs/README.md
Comment (AndersonTorres) [inline pkgs/README.md:26]:
```suggestion
Each new package has a cost for the maintainers, Continuous Integration, caching servers and users downloading
```

More generic and user'ish.

File: pkgs/README.md
Comment (7c6f434c) [reply pkgs/README.md:28]: The precedent is that we don't do much deference to upstream, though. (As in, multipled upstreams tried to stop people from packaging their announced-as-stable OSI-and-DSFG-approved-license software, and Nixpkgs ignored the requests)

Comment (Foxboron) [reply pkgs/README.md:28]: There is a difference between a maintainer going "this isn't stable software and I would urge you to wait" vs "I dont want my FOSS code published in a software repository".

I would ignore the latter if I didn't care about downstream support when encountering bugs.

**Review by picnoir — APPROVED** [review]

File: pkgs/README.md
Comment (infinisil) [reply pkgs/README.md:85]: This document is still a bit of a mess, there's also this section on meta which is probably a better place to change: https://github.com/NixOS/nixpkgs/blob/c5d067d0f23a3412312f036a315bd3e2a8970e9b/pkgs/README.md#meta-attributes

File: pkgs/README.md
Comment (zimbatm) [reply pkgs/README.md]: Added some wording to make the minimum duration normative but not enforced: a0378d09042d3cc4ed5166d20c7f0f9f27913511

File: pkgs/README.md
Comment (zimbatm) [reply pkgs/README.md:85]: I added the license question in the filters: d7b1941ca8c50ff500474bb864b634e921b41dd7

Agreed, once this is merged, I think we should re-format the doc with a focus on the contributor experience, and how they access that information.

File: pkgs/README.md
Comment (zimbatm) [reply pkgs/README.md]: Ok, I addressed this in d5f8b87b2ece652c722a0091c793465e005f6406 . I tried compressing the decision tree so it doesn't become a full D&D adventure.

File: pkgs/README.md
Comment (zimbatm) [reply pkgs/README.md:35]: fixed in 7fae21eb99c81c2f3998db54509c6c81f6063003

**Comment by zimbatm** [comment]
Alright, I believe I have addressed all of the points. The only remaining one is the tonality point by AndersonTorres which I don't really know how to solve.

**Comment by Foxboron** [comment]
I'm no nixpkgs contributor, but speaking from my years as Arch maintainer.

The tone is fine. Adding packages is "easy", it adds additional complexity for the distribution as a whole and removing them is "hard".

The bar to include something should always be there and people adding new packages *should* ask them self "is this only for me, or is this actually useful" before including new things.

If you add poorly maintained projects, someone down the line is going to be want to support newer compiler/dependency targets and be stuck on the poorly maintained project. This work requires effort that not everyone is.

Having a repository of all software in the world, with a complete dependency chain is an impossible job to maintain with an ever increasing amount of complexity.

**Comment by AndersonTorres** [comment]
> The bar to include something should always be there and people adding new packages _should_ ask them self "is this only for me, or is this actually useful" before including new things.
> 

Understandable. 
On the other hand, to cite a personal example, I was the only one interested in packaging _yet another clunky Mplayer fork_. Certainly `mpv` is useful for many people, indeed there are many other people that keep it along with me now. However in the start it was merely another video player, useful only for me.

> This work requires effort that not everyone is capable of committing.

Then it should be about _accountability_. The person should be committed to keep the package first and foremost.

**Comment by zimbatm** [comment]
There is also a catch-22 if we ask packages to be popular before they can get packaged. That's why I didn't want to write those down as hard rules.

Ultimately it's a negotiation that you have to do with yourself. What matters to me is that (1) those points are considered when doing so. And (2) that they can be used in the conversation if somebody else has a different viewpoint. That more or less we have a general understanding of what is valid to package.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/README.md
Comment (AndersonTorres) [inline pkgs/README.md:26]:
```suggestion
We welcome new contributors of new packages to Nixpkgs, arguably the greatest software database known. However, each new package comes with a cost for the maintainers, Continuous Integration, caching servers and users downloading Nixpkgs.
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/README.md
Comment (zimbatm) [reply pkgs/README.md:30]: These points seem close enough that they can be handled by that single point. Let's keep the list as short as possible.

**Comment by RaitoBezarius** [comment]
> On the other hand, to cite a personal example, I was the only one interested in packaging _yet another clunky Mplayer fork_. Certainly `mpv` is useful for many people, indeed there are many other people that keep it along with me now. However in the start it was merely another video player, useful only for me.

This is not a big deal in the NixOS ecosystem, you can start as a NUR, have people be interested in it, promote it to nixpkgs when you figure out a lot of folks are interested in.

Sure, we suck at package discovery across the whole universe with NURs…

File: pkgs/README.md
Comment (SuperSandro2000) [inline pkgs/README.md:36]:
```suggestion
If any of these questions' answer is no, then you should probably not add the package.
```
big fan of one sentence per line like the rest of the file :)

File: pkgs/README.md
Comment (SuperSandro2000) [inline pkgs/README.md:40]:
```suggestion
Luckily it's pretty easy to maintain your own package set with Nix, which can then be added to the [Nix User Repository](https://github.com/nix-community/nur) project.
```

File: pkgs/README.md
Comment (AndersonTorres) [inline pkgs/README.md:30]: That being said, a better wording:

```suggestion
* Is the package ready for general use? We don't want to include projects that are too imature or are going to be abandoned immediately. In case of doubt, check with upstream.
```

**Review by AndersonTorres — COMMENTED** [review]
A non-blocking suggestion:

**Review by AndersonTorres — APPROVED** [review]

**Review by AndersonTorres — APPROVED** [review]
_Nihil Obstat_.

**Review by infinisil — APPROVED** [review]
I have some cleanups for `pkgs/README.md` planned to make it similarly structured to `CONTRIBUTING.md`, with separate "How To's" (where the current Quick Start section should go) and "Conventions" sections (where the new part from this PR should go). But it's not necessary to use that new proposed structure here already, it can be moved later :)

So this looks good!

File: pkgs/README.md
Comment (Artturin) [inline pkgs/README.md:32]: The list seems to be missing a bullet point about novelty programs and programs with multiple implementations in nixpkgs already

See https://github.com/NixOS/nixpkgs/pull/258479 which adds a 4th implementation of lolcat

**Pull request merged**
